x-vector/src/X/Data/Vector/Generic.hs

```diff
@@ -29,9 +44,71 @@ import           Data.Vector.Fusion.Stream.Monadic (Stream(..))
 import qualified Data.Vector.Fusion.Stream.Monadic as Stream
 
 import           Data.Vector.Generic as Generic
+import qualified Data.Vector.Generic.Mutable as MGeneric
+import qualified Data.Vector.Unboxed as Unboxed
 
 import           P hiding (mapMaybe)
 
+-- | The 'transpose' function transposes the rows and columns of its argument.
+--
+--   For example:
+--
+--   > transpose [[1,2,3],[4,5,6]] == [[1,4],[2,5],[3,6]]
+--
+--   If some of the rows are shorter than the following rows, their elements are skipped:
+--
+--   > transpose [[10,11],[20],[],[30,31,32]] == [[10,20,30],[11,31],[32]]
+--
+transpose :: forall vv va a. (Vector va a, Vector vv (va a)) => vv (va a) -> vv (va a)
+transpose xss =
+  Generic.create $ do
+    let
+      xss_len :: Int
+      xss_len =
+        Generic.length xss
+
+      xss_lens :: Unboxed.Vector Int
+      xss_lens =
+        lengths xss
+
+      max_len :: Int
+      max_len =
+        Generic.foldl' max 0 xss_lens
+
+    yss <- MGeneric.unsafeNew max_len
+    for_ [0..max_len-1] $ \i ->
+
+      let
+         loop :: Mutable va s a -> Int -> Int -> Int -> ST s Int
+         loop ys0 !to from xs_len =
+           if i < xs_len then do
+             let
+               xs = xss `Generic.unsafeIndex` from
+               x  = xs  `Generic.unsafeIndex` i
+
+             MGeneric.unsafeWrite ys0 to x
+             pure $! to + 1
+           else
+             pure to
+
+         ys :: va a
+         ys =
+           Generic.create $ do
+             ys0 <- MGeneric.unsafeNew xss_len
+             n <- Generic.ifoldM (loop ys0) 0 xss_lens
+             pure $ MGeneric.unsafeSlice 0 n ys0
```

Comment:
This is where we slice the row down to the actual number of columns required `n`